### DISCUSSION TITLE
icon-offset from feature property

### DISCUSSION THREAD

**Discussion by jarkkoskyttala** [maplibre/maplibre-gl-js#3501] [Q&A]
Hi,

We want to set iconOffset individually for each feature in our source and we calculate that value to be included in the feature "properties", so our source looks something like this:
```
{
            type: 'Feature',
            geometry: {
              type: 'Point',
              coordinates,
            },
            properties: {
             iconOffset
            }
}
```
and then we use it in layer layout with the following `'icon-offset': ['get', 'iconOffset']`

Now begins the problem: This seems to be working fine, the offset is applied correctly but our console gets filled with `Expected value to be of type array<number, 2>, but found string instead.` Im guessing this is because the data is stored as strings in properties. I have tried every variant of expressions but no luck converting it back to number pair.

Any suggestions?

**Comment by neodescis** [comment]
Have you tried casting the property to a number, e.g.:

`'icon-offset': ['number', ['get', 'iconOffset']]`

**Comment by jarkkoskyttala** [comment]
The icon-offset is actually of type [number, number], but i have tried separating x and y to own properties:

```
'icon-offset': [
        'literal',
        [
          ['number', ['get', 'iconOffsetX']],
          ['number', ['get', 'iconOffsetY']],
        ],
      ],
```

but then I get error `Expected array<number, 2> but found array<value, 2> instead` no matter what I do 'get' seems to cast the type to "value"

**Comment by neodescis** [comment]
Ah, I believe `literal` is to be used only with actual JavaScript values/arrays, and cannot be combined with `get` in that way.

In your original question, MapLibre is indeed complaining that the iconOffset property is a string. Are you able to change the data so that it is a `[number, number]` array instead of a string?

**Comment by neodescis** [comment]
I found this old mapbox issue that looks very pertinent, but it seems the problem was never solved...

https://github.com/mapbox/mapbox-gl-js/issues/6155

**Comment by jarkkoskyttala** [comment]
I made reproduction of this issue to codepen and actually it seems that its queryRenderedFeatures that triggers the warnings. If i just make a layer use iconOffset maplibre is perfectly happy until I try to query that feature from map. https://codepen.io/jarkkosky/pen/XWOvQVZ

**Comment by jarkkoskyttala** [comment]
I think I found the answer to my problem in documentation of queryRenderedFeatures (https://maplibre.org/maplibre-gl-js/docs/API/classes/maplibregl.Map/#queryrenderedfeatures)

```The properties value of each returned feature object contains the properties of its source feature. For GeoJSON sources, only string and numeric property values are supported (i.e. null, Array, and Object values are not supported).```

I still think this validation when retrieving the rendered features seems odd and unneccesary but as its documented behavior I guess I will try to think of another solution. If I use `querySourceFeatures`, there is no such validation but I am unable to use that in all cases

**Comment by HarelM** [comment]
Strange behavior indeed...

**Comment by jarkkoskyttala** [comment]
What do you guys think should I open an issue about this?

**Comment by HarelM** [comment]
Sure, it won't hurt to have an issue open, but if you truly want this fixed your best approach would be to create a PR to solve this.

**Comment by jarkkoskyttala** [comment]
I created an issue https://github.com/maplibre/maplibre-gl-js/issues/3502 if someone has insight on how this should be fixed I can try to do a PR possibly
